core/templates/dev/head/pages/email_dashboard/EmailDashboardServices.js

```diff
@@ -0,0 +1,120 @@
+    var queries = [];
+    // Index of currently-shown page of query results.
+    var currentPageIndex = -1;
+
+    var fetchQueriesPage = function(pageSize, cursor) {
+      return $http.get(QUERY_DATA_URL, {
+        params: {
+          num_of_queries_per_page: pageSize,
+          cursor: cursor
+        }
+      }).then(function(response) {
+        return response.data;
+      });
+    };
+
+    return {
+      getQueries: function() {
+        return queries;
+      },
+
+      getCurrentPageIndex: function() {
+        return currentPageIndex;
+      },
+
+      getLatestCursor: function() {
+        return latestCursor;
+      },
+
+      submitQuery: function(data, callback) {
+        var startQueryIndex = (currentPageIndex) * QUERIES_PER_PAGE;
+        var endQueryIndex = (currentPageIndex + 1) * QUERIES_PER_PAGE;
+
+        $http.post(QUERY_DATA_URL, {
+          data: data
+        }).success(function() {
+          // Wait for 500ms and then fetch details of newly submitted query.
+          $timeout(function() {
+            fetchQueriesPage(1, null).then(function(data) {
+              var newQueries = [data.recent_queries[0]];
+              queries = newQueries.concat(queries);
+              callback(queries.slice(startQueryIndex, endQueryIndex));
+            });
+          }, 500);
+        });
+      },
+
+      getNextQueries: function(callback) {
+        var startQueryIndex = (currentPageIndex + 1) * QUERIES_PER_PAGE;
+        var endQueryIndex = (currentPageIndex + 2) * QUERIES_PER_PAGE;
+        currentPageIndex = currentPageIndex + 1;
+
+        if (queries.length > endQueryIndex) {
+          callback(queries.slice(startQueryIndex, endQueryIndex));
+        } else {
+          fetchQueriesPage(QUERIES_PER_PAGE, latestCursor).then(function(data) {
+            queries = queries.concat(data.recent_queries);
+            latestCursor = data.cursor;
+            callback(queries.slice(startQueryIndex, endQueryIndex));
+          });
+        }
+      },
+
+      getPreviousQueries: function() {
+        var startQueryIndex = (currentPageIndex - 1) * QUERIES_PER_PAGE;
+        var endQueryIndex = (currentPageIndex) * QUERIES_PER_PAGE;
+        currentPageIndex = currentPageIndex - 1;
+        return queries.slice(startQueryIndex, endQueryIndex);
+      },
+
+      isNextPageAvailable: function() {
+        var nextQueryIndex = (currentPageIndex + 1) * QUERIES_PER_PAGE;
+        return !(queries.length <= nextQueryIndex) || Boolean(latestCursor);
+      },
+
+      isPreviousPageAvailable: function() {
+        return (currentPageIndex > 0);
+      },
+
+      recheckQueryStatus: function(index, callback) {
+        var queryIndex = (currentPageIndex) * QUERIES_PER_PAGE + index;
+        $http.get(QUERY_STATUS_CHECK_URL, {
+          params: {
+            query_id: queries[queryIndex].id
+          }
+        }).then(function(response) {
+          var data = response.data;
+          queries[queryIndex] = data.query;
```

Comment:
Hm, the index may not be stable (e.g. a new query may be inserted into the front of the list in the meantime). I suggest iterating over the queries until you find the one with the same query id and then replacing that; this way, you can be confident that you're replacing the correct thing. We've run into problems like this for e.g. multiple choice question options.